File: April/Abiji-2020/Golang /Add One Row to tree/add one row to tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func solve(t *TreeNode, val, depth int) {
    if t == nil {
        return
    }
    
    if depth == 2 {
        temp := &TreeNode{Val: val}
        temp.Left = t.Left
        t.Left = temp
        
        temp2 := &TreeNode{Val: val}
        temp2.Right = t.Right
        t.Right = temp2
        
        return
    }
    
    solve(t.Left, val, depth-1)
    solve(t.Right, val, depth-1)
}

func addOneRow(root *TreeNode, val, depth int) *TreeNode {
    if depth == 1 {
        t := &TreeNode{Val: val}
        t.Left = root
        return t
    }
    
    solve(root, val, depth)
    return root
}

File: April/Abiji-2020/C++/Add One Row to Tree/add one row to tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    void addnew(TreeNode* root, int val, int depth){
        if(root==nullptr) return;
       
        if(depth==2){
            
               TreeNode* rightNode = new TreeNode(val,nullptr,root->right);
               TreeNode* leftNode = new TreeNode(val, root->left,nullptr);
               root->right = rightNode;
               root->left = leftNode; 
            
            return;
        }
        addnew(root->left,val,depth-1);
        addnew(root->right,val,depth-1);
    }
public:
    TreeNode* addOneRow(TreeNode* root, int val, int depth) {
         if(depth==1){
            TreeNode* base = new TreeNode(val,root,nullptr);
            return base;
        }
        addnew(root,val,depth);
        return root;
    }
};
